pkg/server/stream/v3/subscription_test.go

```diff
@@ -0,0 +1,60 @@
+package stream
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSotwSubscriptions(t *testing.T) {
+	t.Run("legacy mode properly handled", func(t *testing.T) {
+		sub := NewSotwSubscription([]string{})
+		assert.True(t, sub.IsWildcard())
+
+		// Requests always set empty in legacy mode
+		sub.SetResourceSubscription([]string{})
+		assert.True(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+
+		// Requests always set empty in legacy mode
+		sub.SetResourceSubscription(nil)
+		assert.True(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+
+		// Set any resource, no longer wildcard
+		sub.SetResourceSubscription([]string{"resource"})
+		assert.False(t, sub.IsWildcard())
+		assert.Equal(t, map[string]struct{}{"resource": {}}, sub.SubscribedResources())
+
+		// No longer watch any resource, should not come back to wildcard as no longer in legacy mode
+		// We end up with a watch to nothing
+		sub.SetResourceSubscription(nil)
+		assert.False(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+	})
+
+	t.Run("new wildcard mode from start", func(t *testing.T) {
+		// A resource is provided so the subscription was created in wildcard
+		sub := NewSotwSubscription([]string{"*"})
+		assert.True(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+
+		// Keep wildcard, no change
+		sub.SetResourceSubscription([]string{"*"})
+		assert.True(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+
+		// Add resource to wildcard
+		sub.SetResourceSubscription([]string{"*", "resource"})
+		assert.True(t, sub.IsWildcard())
+		assert.Equal(t, map[string]struct{}{"resource": {}}, sub.SubscribedResources())
+	})
+
+	t.Run("new wildcard mode from start", func(t *testing.T) {
+		sub := NewSotwSubscription([]string{"*"})
+		// Requests always set empty in legacy mode
+		sub.SetResourceSubscription([]string{"*"})
+		assert.True(t, sub.IsWildcard())
+		assert.Empty(t, sub.SubscribedResources())
+	})
```

Comment:
This seems redundant with the previous test, and has the same name. Did you intend to test something else? I think there is more to test here, such as cases where there are both resources by name and wildcards, especially the unsubscription from wildcard case.

Reply:
Yes sorry this part of tests is still in progress. Ended up spending far more time than I expected for the other tests that were mixing requests and streamstates/subscriptions not matching

Reply:
Finalized those tests